### ISSUE TITLE
CentOS 7 - CVE-2019-20372 - nginx 1.18.x (latest stable) + Passenger 6.0.5 or 6.0.6 Dynamic Module Cannot be Installed

### ISSUE THREAD

**Issue by dave105010** [phusion/passenger#2297]
Hi everyone,

I upgraded our passenger to latest version with 
```
sudo yum remove -y nginx nginx-filesystem
sudo yum clean all && sudo yum update
sudo yum install nginx-mod-http-passenger
```
as it was mentioned in the changelog for CentOS 7 users. (https://blog.phusion.nl/2020/05/29/passenger-6-0-5/)

This fetches nginx from EPEL repo as intended but the EPEL nginx version is way too old. (1.16.x)
Before, we had 1.17.x + Passenger 6.0.4
Now, we have 1.16.x + Passenger 6.0.5. 

The latest mainline is 1.19.x and the latest stable is 1.18.x. 

This whole EPEL repo downgrading the nginx to update passenger to 6.0.5 or 6.0.6 with this method caused our security department to warn us regarding to CVEs. 
Therefore, we are forced to update to 1.18.x due to the **CVEs mentioned** below.

https://nvd.nist.gov/vuln/detail/CVE-2019-20372
https://access.redhat.com/security/cve/cve-2019-20372

This a major issue.

> NGINX before 1.17.7, with certain error_page configurations, allows HTTP request smuggling, as demonstrated by the ability of an attacker to read unauthorized web pages in environments where NGINX is being fronted by a load balancer.

However, we cannot update nginx to 1.18.0 due to the passenger dynamic module blocking the upgrade.
Will you guys release an updated passenger binary module to support the latest stable nginx-1.18.x on CentOS 7 soon?
Otherwise, this risks everyone's server environment.

```
Resolving Dependencies
--> Running transaction check
---> Package nginx.x86_64 1:1.16.1-1.el7 will be updated
--> Processing Dependency: nginx = 1:1.16.1 for package: nginx-mod-http-passenger-6.0.5-1.el7.x86_64
---> Package nginx.x86_64 1:1.18.0-1.el7.ngx will be an update
--> Finished Dependency Resolution
Error: Package: nginx-mod-http-passenger-6.0.5-1.el7.x86_64 (@passenger)
           Requires: nginx = 1:1.16.1
           Removing: 1:nginx-1.16.1-1.el7.x86_64 (@epel)
               nginx = 1:1.16.1-1.el7
           Updated By: 1:nginx-1.18.0-1.el7.ngx.x86_64 (nginx)
               nginx = 1:1.18.0-1.el7.ngx
           Available: 1:nginx-1.8.0-1.el7.ngx.x86_64 (nginx)
               nginx = 1:1.8.0-1.el7.ngx
           Available: 1:nginx-1.8.1-1.el7.ngx.x86_64 (nginx)
               nginx = 1:1.8.1-1.el7.ngx
           Available: 1:nginx-1.10.0-1.el7.ngx.x86_64 (nginx)
               nginx = 1:1.10.0-1.el7.ngx
           Available: 1:nginx-1.10.1-1.el7.ngx.x86_64 (nginx)
               nginx = 1:1.10.1-1.el7.ngx
           Available: 1:nginx-1.10.2-1.el7.ngx.x86_64 (nginx)
               nginx = 1:1.10.2-1.el7.ngx
           Available: 1:nginx-1.10.3-1.el7.ngx.x86_64 (nginx)
               nginx = 1:1.10.3-1.el7.ngx
           Available: 1:nginx-1.12.0-1.el7.ngx.x86_64 (nginx)
               nginx = 1:1.12.0-1.el7.ngx
           Available: 1:nginx-1.12.1-1.el7.ngx.x86_64 (nginx)
               nginx = 1:1.12.1-1.el7.ngx
           Available: 1:nginx-1.12.2-1.el7_4.ngx.x86_64 (nginx)
               nginx = 1:1.12.2-1.el7_4.ngx
           Available: 1:nginx-1.14.0-1.el7_4.ngx.x86_64 (nginx)
               nginx = 1:1.14.0-1.el7_4.ngx
           Available: 1:nginx-1.14.1-1.el7_4.ngx.x86_64 (nginx)
               nginx = 1:1.14.1-1.el7_4.ngx
           Available: 1:nginx-1.14.2-1.el7_4.ngx.x86_64 (nginx)
               nginx = 1:1.14.2-1.el7_4.ngx
           Available: 1:nginx-1.16.0-1.el7.ngx.x86_64 (nginx)
               nginx = 1:1.16.0-1.el7.ngx
           Available: 1:nginx-1.16.1-1.el7.ngx.x86_64 (nginx)
               nginx = 1:1.16.1-1.el7.ngx
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest
```

**Comment by CamJN** [comment]
If you want to use another nginx that's fine. There are instructions here: https://www.phusionpassenger.com/docs/advanced_guides/install_and_upgrade/nginx/install_as_nginx_module.html#dynamic-module

**Comment by dave105010** [comment]
> If you want to use another nginx that's fine. There are instructions here: https://www.phusionpassenger.com/docs/advanced_guides/install_and_upgrade/nginx/install_as_nginx_module.html#dynamic-module

Wow, the issue is closed fast without even giving a thought on the whole situation. 
Compiling an entire source code is not something that every system administrator would want to go through. 
Especially, automating this would be very painful for everyone who uses passenger on CentOS.

**Comment by CamJN** [comment]
It's not that no thought has gone into the decision, but that it was considered back when we decided to stop packaging our own Nginx. If you have concerns with the security of the Nginx package that EPEL provides, then that is something to raise in their issue tracker. And as I said above, we provide a workflow for not using the EPEL packaged Nginx, I'm sorry if it isn't to your liking, but it is what we offer for custom configurations.

**Comment by dave105010** [comment]
I already raised that in their issue tracker. Then, I even e-mailed the nginx EPEL package maintainers. I got absolutely no reply. It's just that no one wants to fix things or keep them updated so that we won't be exposed to security vulnerabilities. :(

Moreover, new version of the nginx is already available but then passenger dynamic module provided is not compatible with the new one. 

So, I assume that it should be passenger who maintains the updated the dynamic module that works with at least the latest stable 1.18.x major version of nginx.

Since, you guys stopped packaging your own nginx. Considering that no one has to deal with compiling the entire nginx...

I assume it should be easier and faster to release newer version of the passenger module since nginx stable repo stays pretty standard like EPEL repo.

Is there any particular reason that you guys preferred EPEL repo with older nginx to base passenger module on instead of the official nginx repos?

Again, I don't mean anything bad or offend you. I'm just trying to understand how everything is handled on your side to reach to a common ground to resolve this problem.

**Comment by CamJN** [comment]
We cannot build against the Nginx repo's packages because they do not publish the source rpms for their packages in their rpm repo (you can see here: http://nginx.org/en/linux_packages.html#sourcepackages that they just point you to a mercurial repo, in which you can manually look for the details of any particular release), and as such building a compatible Passenger is a manual process described in the link I sent you, which is not appropriate for releasing Passenger which needs to be automated. Moreover we already require people to add the EPEL repo to have access to certain ruby gem packages, so we felt that was a better choice.

If you are interested in automating the process you could take a look at https://github.com/phusion/passenger_rpm_automation and see if you can find a way to scrape the information you need to build an rpm package to match the nginx repo's packages, however we found it wasn't a reasonable approach when we looked into it.

**Comment by dvershinin** [comment]
@unkn0wn-developer If I may chime in here and you're not against a _commercial_ solution, the dynamic Passenger module is available through the GetPageSpeed repository for CentOS/RHEL 6-7 [example, for CentOS 7 here](https://extras.getpagespeed.com/redhat/7/x86_64/repoview/nginx-module-passenger.html).

As this dynamic NGINX module is built against the latest *official* stable NGINX (not EPEL's) & Passenger *6*, you will also require installing Passenger 6 from the Phusion RPM repo.

1. Enable Phusion Passenger repo.
2. Install Passenger: `sudo yum install -y passenger passenger-devel`
3. Disable Phusion Passenger repo.
4. `sudo yum -y install nginx nginx-module-passenger` (and optionally add any optional dynamic module packages like `nginx-module-pagespeed`)

Result:

* You get to install everything through packages
* Passenger module is dynamic, thus 
* ... you can add any other [50+ NGINX module packages](https://extras.getpagespeed.com/redhat/7/x86_64/repoview/nginx-extras.group.html) by simply installing them.
* No issues with SELinux, etc.
* New NGINX released - the Passenger module is rebuilt and available for update

There are builds for latest mainline NGINX as well, [example, here](https://extras.getpagespeed.com/redhat/7/mainline/x86_64/repoview/nginx-module-passenger.html).

@CamJN 
> they do not publish the source rpms for their packages in their rpm repo

They actually do. See [here](http://nginx.org/packages/centos/7/SRPMS/).

**Comment by CamJN** [comment]
@dvershinin thanks for that, I wonder why they claim not to then. I also thought I tested it at the time... oh well. Hopefully @unkn0wn-developer is happy with their options now.

**Comment by dave105010** [comment]
> @dvershinin thanks for that, I wonder why they claim not to then. I also thought I tested it at the time... oh well. Hopefully @unkn0wn-developer is happy with their options now.

Hey everyone @CamJN @dvershinin ,

That works! 😁  (Took 10-20 minutes to get a working compiled .so file)

Here are the steps that I've taken to compile just the module (not the entire nginx): 

1. yum install nginx [# this installed nginx-1.18.0 on my system from official nginx repos](https://docs.nginx.com/nginx/admin-guide/installing-nginx/installing-nginx-open-source/#installing-a-prebuilt-centos-rhel-package-from-the-official-nginx-repository)
2. curl http://nginx.org/packages/centos/7/SRPMS/nginx-1.18.0-1.el7.ngx.src.rpm --output nginx-1.18.0-source.src.rpm
3. rpm2cpio nginx-1.18.0-source.src.rpm | cpio -idmv # This gave me the source code in tar.gz archive file
4. tar -xvf nginx-1.18.0.tar.gz 
5. cd nginx-1.18.0
6. I did 'nginx -V' and it listed tons of configure arguments. I copied all of them.
7. Then, I did `./configure (all the arguments given with the command above) --add-dynamic-module=$(passenger-config --nginx-addon-dir)` which ended up something like 
```
./configure --prefix=/etc/nginx --sbin-path=/usr/sbin/nginx --modules-path=/usr/lib64/nginx/modules --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --http-log-path=/var/log/nginx/access.log --pid-path=/var/run/nginx.pid --lock-path=/var/run/nginx.lock --http-client-body-temp-path=/var/cache/nginx/client_temp --http-proxy-temp-path=/var/cache/nginx/proxy_temp --http-fastcgi-temp-path=/var/cache/nginx/fastcgi_temp --http-uwsgi-temp-path=/var/cache/nginx/uwsgi_temp --http-scgi-temp-path=/var/cache/nginx/scgi_temp --user=nginx --group=nginx --with-compat --with-file-aio --with-threads --with-http_addition_module --with-http_auth_request_module --with-http_dav_module --with-http_flv_module --with-http_gunzip_module --with-http_gzip_static_module --with-http_mp4_module --with-http_random_index_module --with-http_realip_module --with-http_secure_link_module --with-http_slice_module --with-http_ssl_module --with-http_stub_status_module --with-http_sub_module --with-http_v2_module --with-mail --with-mail_ssl_module --with-stream --with-stream_realip_module --with-stream_ssl_module --with-stream_ssl_preread_module --with-cc-opt='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -fPIC' --with-ld-opt='-Wl,-z,relro -Wl,-z,now -pie' --add-dynamic-module=$(passenger-config --nginx-addon-dir)
```
8. Then, I did `make modules`.
9. It compiled and created `ngx_http_passenger_module.so` file in `obj/` directory.
10. I copied that file to `/etc/nginx/modules`
11. I added `load_module modules/ngx_http_passenger_module.so;` this line into `/etc/nginx/nginx.conf`
12. Then, restarted nginx.

Bam, it worked.

However, I'm not sure what else to do to cover edge-case scenarios. I don't want this to crash or create some binary related errors on production.
I just compiled the passenger module. That's it. I haven't compiled nginx from the source. I prefer the official `nginx` binary from nginx repos to not take anymore responsibility related to its configuration. I have no idea what every ./configure argument does and learning and trying to apply that would take forever and I don't have that much time.

-----

@CamJN @dvershinin  Anyway, if passenger module didn't match with the nginx configure commands, what can potentially happen? Would it warn me with something like `.so binary not compatible with nginx`? or would it just run the module and fail with an unexpected segmentation fault in the middle of the day? (We can't let that happen on production. It would affect a lot of people.)

**Comment by CamJN** [comment]
@unkn0wn-developer yes if the module was incompatible nginx would warn you and refuse to start. Copying the configure arguments from `nginx -V` is often good enough, but not always, which is why that's not the process we document. In this case though if nginx loads the module then it worked just fine.

**Comment by dvershinin** [comment]
@unkn0wn-developer my main point was that if you opt to use the GetPageSpeed repository, you don't have to compile neither NGINX nor the Passenger module and get around with just using `yum`, which is the best scenario for the production use case.

If you're not into using a commercial repository (GetPageSpeed is a paid repo), it is best to create your own RPM package for the dynamic Passenger NGINX module. NGINX blog has a thorough [walkthrough](https://www.nginx.com/blog/creating-installable-packages-dynamic-modules/) for it.

The main downsides to "compile module only" approach is potentially broken NGINX instance if you update `nginx` package by accident (the packaged NGINX and compiled module will not match each other properly). So you may want to `yum versionlock nginx` to prevent unwanted NGINX upgrades. The other issue is compiling module on a production system. Compilation software is a security risk on its own, so the recommendation there is compiling the module outside of the live system.

**Comment by dave105010** [comment]
@CamJN @dvershinin Thank you both. I currently created the nginx passenger module on my dev. I don't think we would want to compile this again and again for every live server. We might either use the same .so file for every live server or think about something else. I'm not sure yet. We might versionlock nginx. I'm not sure about GetPageSpeed as well. It's a third-party provider. I'm not sure if we can trust that. 

We prefer to get the packages from the official sources to prevent security risks. 
So, I'm still stuck because I don't think my current solution is an acceptable way to put this in production even though it works on my dev.

### **One more important thing**
I'm not sure what @CamJN thinks about the current possibility of having an updated the passenger module that works with 1.18.0 after the new findings. 
But... I also opened **RedHat Bugzilla ticket for EPEL repo** for them to update the nginx: https://bugzilla.redhat.com/show_bug.cgi?id=1867261
If you guys can chime in with a comment or something 😀, it would help to get their attention to get the nginx updated so that this can be resolved for everyone with less work required by passenger peeps 😉 

Someone else [opened this old redhat bugzilla ticket](https://bugzilla.redhat.com/show_bug.cgi?id=1750857) last year to get the nginx updated for EPEL repos before. This resulted with them updating it. I took the same steps with my new ticket. Hopefully EPEL maintainers will update it to the latest stable nginx 🤞  *fingers crossed*
 
@FooBarWidget
